core/src/test/java/com/linecorp/armeria/client/DnsMetricsTest.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.ExecutionException;
+
+import org.junit.jupiter.api.Test;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.metric.MoreMeters;
+import com.linecorp.armeria.common.metric.PrometheusMeterRegistries;
+
+import io.micrometer.prometheus.PrometheusMeterRegistry;
+
+public class DnsMetricsTest {
+
+    @Test
+    void dns_metric_test_for_successful_query_writes() throws ExecutionException, InterruptedException {
+        final ClientFactory factory = ClientFactory.builder()
+                .meterRegistry(PrometheusMeterRegistries.newRegistry())
+                .build();
+
+        final WebClient client2 = WebClient.builder()
+                .factory(factory)
+                .build();
+
+        client2.execute(RequestHeaders.of(HttpMethod.GET, "http://wikipedia.com")).aggregate().get();
+
+        final double count = ((PrometheusMeterRegistry) factory.meterRegistry())
+                .getPrometheusRegistry()
+                .getSampleValue("armeria_client_dns_queries_total",
+                        new String[] {"cause","name","result"},
+                        new String[] {"","wikipedia.com.", "success"});
+        assertThat(count > 1.0).isTrue();
+    }
+
+    @Test
+    void dns_metric_test_for_query_failures() throws ExecutionException, InterruptedException {
+        final ClientFactory factory = ClientFactory.builder()
+                .meterRegistry(PrometheusMeterRegistries.newRegistry())
+                .build();
+        try {
+            final WebClient client2 = WebClient.builder()
+                    .factory(factory)
+                    .build();
+            client2.execute(RequestHeaders.of(HttpMethod.GET, "http://googleusercontent.com")).aggregate().get();
+        } catch (Exception ex) {
+            final double count = ((PrometheusMeterRegistry) factory.meterRegistry())
+                    .getPrometheusRegistry()
+                    .getSampleValue("armeria_client_dns_queries_total",
+                            new String[] {"cause","name","result"},
+                            new String[] {"No matching record type found","googleusercontent.com.", "failure"});
+            assertThat(count > 1.0).isTrue();
+        }
+    }
+
+    @Test
+    void no_dns_registry_used_when_not_provided_externally() throws ExecutionException, InterruptedException {
+        final ClientFactory factory = ClientFactory.builder()
+                .build();
```

Comment:
Global comment: `ClientFactory` must be closed at the end of the test. Use `try-with-resources` on it.